src/main/java/com/salesforce/dataloader/client/HttpClientTransport.java

```diff
@@ -90,64 +95,83 @@ public OutputStream connect(String url, String soapAction) throws IOException {
+                credentialsprovider.setCredentials(scope, credentials);
+                client = HttpClientBuilder.create().setDefaultCredentialsProvider(credentialsprovider).build();
             }
-            
-            // copy input stream data into a new input stream because releasing the connection will close the input stream
-            ByteArrayOutputStream bOut = new ByteArrayOutputStream();
-            IOUtils.copy(response.getEntity().getContent(), bOut);
-            input = new ByteArrayInputStream(bOut.toByteArray());
-
-            if (response.containsHeader("Content-Encoding") && response.getHeaders("Content-Encoding")[0].getValue().equals("gzip")) {
-                input = new GZIPInputStream(input);
+
+
+
+            byte[] entityBytes = entityByteOut.toByteArray();
+            HttpEntity entity = new ByteArrayEntity(entityBytes);
+            post.setEntity(entity);
+
+            CloseableHttpResponse response = null;
+            try {
+                if (config.getNtlmDomain() != null && !config.getNtlmDomain().equals("")) {
+                    // need to send a HEAD request to trigger NTLM authentication
+                    HttpHead head = new HttpHead("http://salesforce.com");
+                    try {
+                        response = client.execute(head);
+                    } finally {
+                        response.close();
+                    }
+                }
+                response = client.execute(post);
+
+                if (response.getStatusLine().getStatusCode() > 399) {
+                    successful = false;
+                    if (response.getStatusLine().getStatusCode() == 407) {
+                        throw new RuntimeException(response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
+                    }
+                } else {
+                    successful = true;
+                }
+
+                // copy input stream data into a new input stream because releasing the connection will close the input stream
+                ByteArrayOutputStream bOut = new ByteArrayOutputStream();
+                InputStream inStream = response.getEntity().getContent();
+                try {
+                    IOUtils.copy(response.getEntity().getContent(), bOut);
+                    input = new ByteArrayInputStream(bOut.toByteArray());
+
+                    if (response.containsHeader("Content-Encoding") && response.getHeaders("Content-Encoding")[0].getValue().equals("gzip")) {
+                        input = new GZIPInputStream(input);
+                    }
+                } finally {
+                    inStream.close();
+                }
+
+            } finally {
+                if (response != null)
+                    response.close();
             }
-   
         } finally {
-            post.releaseConnection();
+            client.close();
```

Comment:
:) Sorry the diffs are kind of hard to follow. Adding Try block messed around it a lot!